src/Interval.h

```diff
@@ -0,0 +1,301 @@
+    }
+
+    /// `center` returns the center of this interval. It is arbitrary
+    /// for empty intervals.
+    Scalar center() const { return 0.5 * (_a + _b); }
+    /// `size` returns the size (length, width) of this interval. It is zero
+    /// for single-point intervals, and NaN or negative for empty intervals.
+    Scalar size() const { return _b - _a; }
+
+    ///@{
+    /// `contains` returns true if the intersection of this interval and x
+    /// is equal to x.
+    bool contains(Scalar x) const {
+        return (_a <= x && x <= _b) || x != x;
+    }
+    bool contains(Interval const & x) const {
+        if (x.isEmpty()) {
+            return true;
+        } else if (isEmpty()) {
+            return false;
+        }
+        return _a <= x._a && _b >= x._b;
+    }
+    ///@}
+
+    ///@{
+    /// `isDisjointFrom` returns true if the intersection of this interval
+    /// and x is empty.
+    bool isDisjointFrom(Scalar x) const {
+        return !intersects(x);
+    }
+    bool isDisjointFrom(Interval const & x) const {
+        if (isEmpty() || x.isEmpty()) {
+            return true;
+        }
+        return _a > x._b || _b < x._a;
+    }
+    ///@}
+
+    ///@{
+    /// `intersects` returns true if the intersection of this interval and x
+    /// is non-empty.
+    bool intersects(Scalar x) const { return _a <= x && x <= _b; }
+    bool intersects(Interval const & x) const {
+        return !isDisjointFrom(x);
+    }
+    ///@}
+
+    ///@{
+    /// `isWithin` returns true if the intersection of this interval and x
+    /// is this interval.
+    bool isWithin(Scalar x) const {
+        return (_a == x && _b == x) || isEmpty();
+    }
+    bool isWithin(Interval const & x) const {
+        return x.contains(*this);
+    }
+    ///@}
+
+    ///@{
+    /// `relate` returns a bit field F describing the spatial relations
+    /// between this interval and x. For each relation that holds, the bitwise
+    /// AND of F and the corresponding `SpatialRelation` bit mask will be
+    /// non-zero.
+    int relate(Scalar x) const;
+    int relate(Interval const & x) const;
+    ///@}
+
+    ///@{
+    /// `shrinkTo` shrinks this interval until all its points are in x.
+    Interval & shrinkTo(Scalar x) {
+        if (x != x) {
+            _a = x;
+            _b = x;
+        } else {
+            _a = std::max(_a, x);
+            _b = std::min(_b, x);
+        }
+        return *this;
+    }
+    Interval & shrinkTo(Interval const & x) {
+        if (x.isEmpty()) {
+            *this = x;
+        } else if (!isEmpty()) {
+            _a = std::max(_a, x._a);
+            _b = std::min(_b, x._b);
+        }
+        return *this;
+    }
+    ///@}
+
+    ///@{
+    /// `shrunkTo` returns the intersection of this interval and x.
+    Derived shrunkTo(Scalar x) const { return Interval(*this).shrinkTo(x); }
+    Derived shrunkTo(Interval const & x) const {
+        return Interval(*this).shrinkTo(x);
+    }
+    ///@}
```

Comment:
Oh, and I guess I should add that `afw::geom::Box` uses `clipTo` for in-place intersection (but doesn't have a version that returns a new object).  It might be good to make that consistent, and go with `clipTo` and `clippedTo` here.